src/full/Agda/TypeChecking/Unquote.hs

```diff
@@ -1041,6 +1044,35 @@ evalTCM v = do
       locallyReduceAllDefs $ checkFunDef NotDelayed i x cs
       primUnitUnit
 
+    tcPragmaForeign :: Text -> Text -> TCM Term
+    tcPragmaForeign backend code = do
+      modifyTC $ \ st -> let preScSt = stPreScopeState st in st
+        { stPreScopeState = preScSt
+          { stPreForeignCode = Map.insertWith
+            (flip (++))
+            (T.unpack backend)
+            [ForeignCode noRange $ T.unpack code]
+            (stPreForeignCode preScSt)
+          }
+        }
+      primUnitUnit
+
+    tcPragmaCompile :: Text -> QName -> Text -> TCM Term
+    tcPragmaCompile backend name code = do
+      modifyTC $ \ st -> let postScSt = stPostScopeState st in st
+        { stPostScopeState = postScSt
+          { stPostSignature = (stPostSignature postScSt)
+            { _sigDefinitions = HashMap.adjust
+              (\ d -> d
+                { defCompiledRep = Map.insertWith (++) (T.unpack backend) [CompilerPragma noRange (T.unpack code)] $ defCompiledRep d
+                })
+              name
+              (_sigDefinitions $ stPostSignature postScSt)
+            }
+          }
+        }
```

Comment:
Looks like you could just compose these two lenses:
https://github.com/agda/agda/blob/318f56b7a4ad2afe06a3242eac80c2a63e9e8044/src/full/Agda/TypeChecking/Monad/Base.hs#L646-L649
https://github.com/agda/agda/blob/318f56b7a4ad2afe06a3242eac80c2a63e9e8044/src/full/Agda/TypeChecking/Monad/Base.hs#L1703-L1706

Reply:
At first I tried, but maybe using the wrong functions (from `Agda.Utils.Lens`), which require `MonadState` while TCM doesn't implement it.

Reply:
> At first I tried, but maybe using the wrong functions (from `Agda.Utils.Lens`), which require `MonadState` while TCM doesn't implement it.

We have `modifyTCLens` and similar to that end. https://github.com/agda/agda/blob/318f56b7a4ad2afe06a3242eac80c2a63e9e8044/src/full/Agda/TypeChecking/Monad/Base.hs#L4719-L4748